### PR TITLE
Further improve and streamline aircraft Fly activity

### PR DESCRIPTION
**Pull request by reaperrr**
Another step towards getting #15803 into bleed.

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/Air/Fly.cs [hunk 1]
```diff
@@ -51,24 +52,38 @@ public Fly(Actor self, Target t, WPos? initialTargetPosition = null, Color? targ
...
-		public static void FlyToward(Actor self, Aircraft aircraft, int desiredFacing, WDist desiredAltitude, int turnSpeedOverride = -1)
+		public static bool FlyToward(Actor self, Aircraft aircraft, int desiredFacing, WDist desiredAltitude, int turnSpeedOverride = -1, bool isVTOL = false)
+		{
+			var move = aircraft.Info.CanHover ? aircraft.FlyStep(desiredFacing) : aircraft.FlyStep(aircraft.Facing);
+			return FlyToward(self, aircraft, move, desiredFacing, desiredAltitude, turnSpeedOverride, isVTOL);
+		}
+
+		// Note: isVTOL is meant to only refer to actual take-off or land, not mere terrain-height-induced altitude changes.
+		// This is to avoid VTOLs constantly 'braking' on terrain height changes.
+		public static bool FlyToward(Actor self, Aircraft aircraft, WVec move, int desiredFacing, WDist desiredAltitude, int turnSpeedOverride = -1, bool isVTOL = false)
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]: Can we please replace `bool isVTOL` with `MovementType types`?

This turns a hack into a feature and creates a relatively coherant external interface.
We can move the facing tick logic inside this method for `MovementType.Turn`.

File: OpenRA.Mods.Common/Activities/Air/Fly.cs [hunk 2]
```diff
@@ -96,38 +111,73 @@ public override Activity Tick(Actor self)
...
+				{
+					Game.Sound.Play(SoundType.World, aircraft.Info.TakeoffSounds, self.World, aircraft.CenterPosition);
+					soundPlayed = true;
+				}
+
 				aircraft.RemoveInfluence();
+				isTakeOff = true;
+			}
+
+			// If we're a VTOL, rise before flying forward
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]: IMO we should allow `CanHover` units to turn while climbing.

With my suggestion above, we could call `FlyTowards` with `MovementType.Vertical | MovementType.Turn` for VTOL+Hover vs just `MovementType.Vertical` for non-hovering (unless we want it there too?)

Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Air/Fly.cs]: > unless we want it there too?

I went for 'yes', as it makes the code a bit simpler, too.

**Comment by reaperrr** [comment]
Updated (and rebased).

**Comment by pchote** [comment]
Needs a rebase now after #16417.

**Comment by reaperrr** [comment]
Rebased and greatly improved last commit.

**Review by pchote — COMMENTED** [review]
* There is something very weird with carryalls in now. After attaching to a unit they take off diagonally, and freeze in the air if they reach the ordered cell before they finish climbing to cruise altitude. They stay frozen on the ground when ordered to unload - visible in both TS and D2k.
* Helicopters will turn while taking off from the ground or service depot, but not when they are first produced from the helipad.

**Comment by reaperrr** [comment]
Rebased, fixed the style issues, moved take-off movement to `TakeOff`.

However, unfortunately I wasn't able to debug, let alone fix the Carryall/Transport take-off bug, no friggin' clue what's going on there. 4 hours spent with zero success, so I pushed what I have as at this rate, I don't think I'll be able to debug this anytime soon myself.

**Comment by reaperrr** [comment]
Since this became rather unwieldy and hard to debug, I've taken a step backwards, started with the first two commits as base, fixed all regressions I ran into, added the `TakeOff` refactor on top and pushed the result to #16584, which this PR will be rebased on later.

**Comment by reaperrr** [comment]
Closing for now, as I want to focus on bugfixes and polishing first, whereas stuff like merging the vertical-only movement into FlyTick by using MovementTypes might cause new regressions, so might not be a good idea until after the playtest branch-off.

Due to that and the fact this would need a rather hefty rebase (likely more sensible to reapply those changes to a 'fresh' bleed-based branch), I might rather file a new PR later instead of reopening this.

**Pull request closed without merging**